### DISCUSSION TITLE
Welcome to gocron Discussions!

### DISCUSSION THREAD

**Discussion by JohnRoesler** [go-co-op/gocron#88] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by NWBY** [comment]
Hey everyone, I'm using gocron in a server monitoring agent that I'm building and really love the package so far. 
However, I'm struggling to use exported functions from different packages as the task passed to `Do()`. What I'm trying to do is declare a function in another package called `core` then use that function in `main` and pass it to the scheduler, however the error that I get is `functionName() used as value`

Something along the lines of: 
```go
package main

import (
  "/module/core"
  "github.com/go-co-op/gocron"
)

func main() {
	sch := gocron.NewScheduler(time.UTC)

	sch.Every(30).Seconds().Do(core.GetCpuUsage())

	sch.StartAsync()
}
```

Any advice, help or guidance would be greatly appreciated. Apologies in advance if this a stupid question.

**Reply by JohnRoesler** [reply]
Glad to hear you like the package! Try

 1. Referencing the function without parens `sch.Every(30).Seconds().Do(core.GetCpuUsage)`
 2. Create a var and assign the function first `var myFunc = core.GetCpuUsage`

**Reply by NWBY** [reply]
@JohnRoesler Thank you so much for your quick response. Option 1 worked a treat.

**Reply by JohnRoesler** [reply]
Awesome. Feel free to reach out with more questions. We're also on slack (in the readme) 😄

**Comment by Chemical-Battery** [comment]
Hello,

I'm a new programmer, and new to Go. I found your package and I think it would be great for the project I am working on. The idea is I want a Go script that will execute a external Go file from CMD I have already made on a 30 minute interval throughout the day. I wanted to build the script into an executable when I was done, so my coworkers could just click it at the start of their shifts and forget about it. I have attempted to build something out, but I can't seem to get it to kick off. I'm sure I'm missing something obvious, but I can't seem to find the solution in the docs. Would you be able to help me out?  

Here is my code so far.  

                package main
                
                import (
	                "fmt"
	                "os/exec"
	                "time"
                
	                "github.com/go-co-op/gocron"
                )
                
                var task = func() {
	                cmd := exec.Command(`python`, `C:\Users\myusername\OneDrive\Desktop\Test_1\hello.py`)
                
	                cmd.Run()
                }
                
                func main() {
	                fmt.Println("Timer starting")
	                s := gocron.NewScheduler(time.UTC)
                
	                _, _ = s.Every(1).Seconds().Do(task)
	                fmt.Println(s.TaskPresent(task))
                
                }

I'm testing it against a external Python Script here.

**Reply by JohnRoesler** [reply]
hi @Chemical-Battery I'm on a mac so I can't help with the windows aspect of this, but this task worked for me:

```golang
var task = func() {
	out, err := exec.Command("python", `test.py`).Output()

	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf(string(out))
}
```

You've go the gocron part right. I suspect something with the python command isn't working properly.

**Reply by Chemical-Battery** [reply]
How would I use your implementation above to run a Go file instead?

**Reply by JohnRoesler** [reply]
If you built the go binary into an executable you should be able to call it directly. This [example](https://www.freecodecamp.org/news/setting-up-go-programming-language-on-windows-f02c8c14e2f/) has some details on setting up Go in windows in case that is helpful. But looking at that, you should be able to call a go program like: `%GOPATH%/bin/hello` or similar. So assuming you have one built, a main file named `hello`, you'd run `go build hello`. And then I tweaked your example from above 

```diff
package main

import (
    "fmt"
    "os/exec"
    "time"

    "github.com/go-co-op/gocron"
)

var task = func() {
+    cmd := exec.Command("C:\Users\myusername\OneDrive\Desktop\Test_1\hello")
-    cmd := exec.Command(`python`, `C:\Users\myusername\OneDrive\Desktop\Test_1\hello.py`)
    cmd.Run()
}

func main() {
    fmt.Println("Timer starting")
    s := gocron.NewScheduler(time.UTC)

    _, _ = s.Every(1).Seconds().Do(task)
    fmt.Println(s.TaskPresent(task))

}
```

**Reply by Chemical-Battery** [reply]
Okay, I'll take a look at that article. I was able to pull of something like this with my first file, and have it run both a executable and a non-executable Python file. I really appreciate you taking the time to help work with me on this. I'll report back with the results.

**Reply by Chemical-Battery** [reply]
Hello,

I'm happy to report that I was able to get the script running, here is the full code.  
```go
package main

go import (
	"fmt"
	"log"
	"os/exec"
	"time"

	"github.com/go-co-op/gocron"
)

var task = func() {

	cmd := exec.Command("go", "run", `C:\Users\username\OneDrive\Desktop\Testbed\TaskSch_v1\TaskSch.go`)
	out, err := cmd.CombinedOutput()
	if err != nil {
		log.Fatalf("cmd.Run() failed with %s\n", err)
	}
	fmt.Printf("combined out:\n%s\n", string(out))

}

func main() {
	task()

	fmt.Println("Timer starting")
	s := gocron.NewScheduler(time.UTC)

	_, _ = s.Every(30).Minutes().Do(task)
	fmt.Println(s.TaskPresent(task))
	s.StartBlocking()

}  
```


I found the right parameters to put into the Command line to run another Go Script. My other main issue was not having the s.StartBlocking() Function to stop Main from exiting the program. It all works great! This package has been really helpful, I hope I can help contribute to it when I get to a higher skill level.  

The other thing I was wondering is if there is some kind of function within Gocron to have this script run while the computer is asleep? Like say having my laptop lid closed. I tested this with the current code and it looked like it stopped running while asleep. But started back up again when I opened the lid and logged in.

**Reply by JohnRoesler** [reply]
Awesome news! Regarding the ability to run while asleep, that wouldn’t be related to this library but would be a matter of how your computer is architected. During sleep, no I wouldn’t expect it would run. You can change your settings so it doesn’t sleep when the lid is closed and then it would still run.

**Reply by Chemical-Battery** [reply]
I feel silly for not realizing that. Okay, then this should work great for my project at work. Thanks again!

**Reply by JohnRoesler** [reply]
No worries! We were all there once!
